src/Response/JsonResponse.php

```diff
@@ -59,16 +69,84 @@ public function __construct(
         array $headers = [],
         $encodingOptions = self::DEFAULT_JSON_FLAGS
     ) {
-        $body = new Stream('php://temp', 'wb+');
-        $body->write($this->jsonEncode($data, $encodingOptions));
-        $body->rewind();
+        $this->setPayload($data);
+        $this->encodingOptions = $encodingOptions;
+
+        $json = $this->jsonEncode($data, $this->encodingOptions);
+        $body = $this->createBodyFromJson($json);
 
         $headers = $this->injectContentType('application/json', $headers);
 
         parent::__construct($body, $status, $headers);
     }
 
     /**
+     * @return mixed
+     */
+    public function getPayload()
+    {
+        return $this->payload;
+    }
+
+    /**
+     * @param $data
+     *
+     * @return JsonResponse
+     */
+    public function withPayload($data)
+    {
+        $new = clone $this;
+
+        $new->setPayload($data);
+
+        $json = $this->jsonEncode($new->payload, $new->encodingOptions);
+        $body = $this->createBodyFromJson($json);
+
+        $new = $new->withBody($body);
+
+        return $new;
+    }
+
+    /**
+     * @return int
+     */
+    public function getEncodingOptions()
+    {
+        return $this->encodingOptions;
+    }
+
+    /**
+     * @param int $encodingOptions
+     *
+     * @return JsonResponse
+     */
+    public function withEncodingOptions($encodingOptions)
+    {
+        $new = clone $this;
+        $new->encodingOptions = $encodingOptions;
+
+        $json = $this->jsonEncode($new->payload, $new->encodingOptions);
+        $body = $this->createBodyFromJson($json);
+        $new = $new->withBody($body);
```

Comment:
I'd extract these three lines into a separate method, in order to prevent duplication. Something along the lines of:

```php
private function updateBodyFor(self $toUpdate)
{
    $json = $this->jsonEncode($toUpdate->payload, $toUpdate->encodingOptions);
    $body = $this->createBodyFromJson($json);
    return $toUpdate->withBody($body);
}
```

Then, update this method and others like it so that they read:

```php
return $this->updateBodyFor($new);
```